_posts/2020-07-29-parallelized-sampling/parallelized-sampling.Rmd

```diff
@@ -0,0 +1,290 @@
+
+`SWOR` is conceptually equivalent to a $n$-step process of selecting 1 out of $(n - j + 1)$ remaining rows in the $j$-th step for $j \in \{1, \dotsc, n\}$, with each remaining row's likelihood of getting selected being linearly proportional to its weight in any of the steps, i.e.,
+
+```
+samples := {}
+population := {r[1], ..., r[N]}
+
+for j = 1 to n
+  select r[x] from population with probability
+    (w[x] / TotalWeight(population))
+  samples := samples + {r[x]}
+  population := population - {r[x]}
+```
+
+Notice the outcome of a `SWoR` process is in fact order-significant, which is why in this post it will always be represented as an ordered tuple of elements.
+
+Intuitively, `SWoR` is analogous to throwing darts at a bunch of tiles. For example, let's say the size of our sample space is 5:
+
+- Imagine $r_1, r_2, \dotsc, r_5$ as 5 rectangular tiles laid out contiguously on a wall with widths $w_1, w_2, \dotsc, w_5$, with $r_1$ covering $[0, w_1)$, $r_2$ covering $[w_1, w_1 + w_2)$, ..., and $r_5$ covering $\left[\sum\limits_{j = 1}^{4} w_j, \sum\limits_{j = 1}^{5} w_j\right)$
+  
+- Equate drawing a random sample in each step to throwing a dart uniformly randomly within the interval covered by all tiles that are not hit yet
+  
+- After a tile is hit, it gets taken out and remaining tiles are re-arranged so that they continue to cover a contiguous interval without overlapping
+
+If our sample size is 3, then we shall ask ourselves what is the probability of the dart hitting $(r_1, r_2, r_3)$ in that order? 
+
+In step $j = 1$, the dart will hit $r_1$ with probability $\left. w_1 \middle/ \left(\sum\limits_{k = 1}^{N}w_k\right) \right.$
+
+![step 1](images/d1.jpg){ width=50% height=50% } .
+
+After deleting $r_1$ from the sample space after it's hit, step $j = 2$ will look like this:
+
+![step 2](images/d2.jpg){ width=48% height=45% } ,
+
+and the probability of the dart hitting $r_2$ in step 2 is $\left. w_2 \middle/ \left(\sum\limits_{k = 2}^{N}w_k\right) \right.$ .
+
+Finally, moving on to step $j = 3$, we have:
+
+![step 3](images/d3.jpg){ width=40% height=30% } ,
+
+with the probability of the dart hitting $r_3$ being $\left. w_3 \middle/ \left(\sum\limits_{k = 3}^{N}w_k\right) \right.$.
+
+So, combining all of the above, the overall probability of selecting  $(r_1, r_2, r_3)$ is $\prod\limits_{j = 1}^{3} \left( {w_j} \middle/ {\sum\limits_{k = j}^{N}{w_k}} \right)$.
+
+## Naive approaches for implementing `SWoR`
+
+This section outlines some possible approaches that were briefly under consideration. Because none of these approaches scales well to a large number of rows or a non-trivial number of partitions in a Spark dataframe, we decided to avoid all of them in `sparklyr`.
```

Comment:
```suggestion
This section outlines some possible approaches that we briefly considered. Because none of these approaches scale well to a large number of rows or a non-trivial number of partitions in a Spark dataframe, we decided to avoid all of them in `sparklyr`.
```

Reply:
Just minimizing passive voice here.